### PR TITLE
[SPARK-10671] [SQL] Throws an analysis exception if we cannot find Hive UDFs

### PR DESCRIPTION
**Pull request by yhuai**
https://issues.apache.org/jira/browse/SPARK-10671

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 1]
```diff
@@ -72,7 +72,8 @@ class DataSourceWithHiveMetastoreCatalogSuite
       "org.apache.hadoop.hive.ql.io.orc.OrcSerde"
     )
   ).foreach { case (provider, (inputFormat, outputFormat, serde)) =>
-    test(s"Persist non-partitioned $provider relation into metastore as managed table") {
+    // TODO: Re-enable it before merging this PR.
+    ignore(s"Persist non-partitioned $provider relation into metastore as managed table") {
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: This change is for testing.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 2]
```diff
@@ -97,7 +98,8 @@ class DataSourceWithHiveMetastoreCatalogSuite
       }
     }
 
-    test(s"Persist non-partitioned $provider relation into metastore as external table") {
+    // TODO: Re-enable it before merging this PR.
+    ignore(s"Persist non-partitioned $provider table into metastore as external table") {
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: This change is for testing.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 3]
```diff
@@ -156,5 +158,20 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
+          val path = dir.getCanonicalPath
+
+          sql(
+            s"""CREATE TABLE t USING $provider
+               |OPTIONS (path '$path')
+               |AS SELECT 1 AS timestamp, cast("val_1" as binary) AS `binary.dot`
+             """.stripMargin)
+        }
+      }
+    }
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: This change is for testing.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 4]
```diff
@@ -46,6 +46,8 @@ object TestHive
       "TestSQLContext",
       new SparkConf()
         .set("spark.sql.test", "")
+        .set("spark.sql.hive.metastore.version", "0.13.1") // TODO: remove this setting.
+        .set("spark.sql.hive.metastore.jars", "maven") // TODO: remove this setting.
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: This change is for testing.

**Comment by liancheng** [comment]
Currently `saveAsTable` tries its best to save the table as Hive compatible format, and falls back to Spark SQL specific format if it fails. I wonder whether it's too confusing for users to reason about behavior of `saveAsTable`. Would it be better to have a separate SQL option to force using Hive compatible format? With this option on, users can be quite confident that if a `saveAsTable` call succeeds, the saved table must be Hive compatible.

**Comment by yhuai** [comment]
The impact of saving the metadata in a hive compatible way or not will appear when a user want to read the data. I am not sure they can remember how the metadata is stored. It is also very possible that the data producer and consumer are different persons/different teams.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 5]
```diff
@@ -59,19 +59,52 @@ private[hive] class HiveFunctionRegistry(underlying: analysis.FunctionRegistry)
...
+        // To trigger any error like arguments are not valid, we call the dataType method
+        // to make sure internal Hive function resolver gets created without any issue.
+        // The only exception at here is GenericUDTF, because it does not support dataType
+        // method, we call elementTypes instead.
+        val resolvedHiveFunction =
+          if (classOf[UDF].isAssignableFrom(functionInfo.getFunctionClass)) {
+            val func = HiveSimpleUDF(new HiveFunctionWrapper(functionClassName), children)
+            // We actually only need func.dataType. Assigning the returned value to a val
+            // is just to make the compiler happy.
+            val notUsed = func.dataType
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: hmm, it is not really safe to call `dataType` to trigger the evaluation of those lazy vals.

**Comment by yhuai** [comment]
Will split this PR to two PRs.

**Comment by davies** [comment]
LGTM

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 6]
```diff
@@ -59,19 +59,34 @@ private[hive] class HiveFunctionRegistry(underlying: analysis.FunctionRegistry)
...
-      } else if (
-        classOf[AbstractGenericUDAFResolver].isAssignableFrom(functionInfo.getFunctionClass)) {
-        HiveGenericUDAF(new HiveFunctionWrapper(functionClassName), children)
-      } else if (classOf[UDAF].isAssignableFrom(functionInfo.getFunctionClass)) {
-        HiveUDAF(new HiveFunctionWrapper(functionClassName), children)
-      } else if (classOf[GenericUDTF].isAssignableFrom(functionInfo.getFunctionClass)) {
-        HiveGenericUDTF(new HiveFunctionWrapper(functionClassName), children)
-      } else {
-        sys.error(s"No handler for udf ${functionInfo.getFunctionClass}")
+      try {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: Should we add a comment here to say that we do type checking inside hive UDF class(such as `HiveSimpleUDF`) constructor?

**Pull request closed without merging**